vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+ * blocks in each partition, and the difference in the utilization (Occupancy / Capacity) of each subarea.
+ *
+ * To generate the source cut, the logic blocks are sorted by their raw_x or raw_y location, depending on the
+ * orientation of the desired cut. After sorting, a pivot is chosen, all blocks to the left of the pivot are
+ * assigned to the left partition, and all blocks to the right are assigned to the right partition (we use left
+ * for left/top, right for right/bottom in x/y directions respectively in the implementation).
+ *
+ * The target cut is an x or y cut of the area such that all blocks in each partition fit in their respective
+ * subareas, and the difference in utilization is minimized. This is difficult due to the discrete nature of FPGA
+ * architecture. To achieve this, first assign all tiles to right region, then move tiles to the left one by one,
+ * calculating the difference in utilization for each move. The move with smallest delta_utilization is chosen as
+ * the target cut. To eliminate possible overutilization in either subarea, perturb the source cut by moving a
+ * single block from the over-utilized sub-area to the other, until neither is overutilized.
+ *
+ * Next, the blocks in sub-areas are spread to distribute them evenly. We split the sub-area into 10 equally-sized
+ * bins and logic blocks into 10 equal-capacity source bins. Then linearly interpolate to map blocks from their
+ * original_locations in their source bins to new spread location in target bins.
+ *
+ * This cut and spreading is repeated recursively. The cut spreading process returns the left and right subareas,
+ * which are pushed into a FIFO. Then a region is popped from the FIFO and goes through cut-spreading. This process
+ * continues until a base case of only 1 block in the region is reached. At this point the placement is mostly not
+ * overutilized and ready for strict legalization.
+ *
+ * **************************************************************************************************************
+ *
+ * Strict Legalizer
+ * ================
+ * @see strict_legalize()
+ * Strict Legalizer ensures that the placement is strictly legal. It does so using a greedy approach.
+ *
+ * All blocks are sorted in descending macro lengths order and put in a priority queue (only macro heads are
+ * considered, while the rest of the macro members are ignored; single blocks have length 1). Each block goes through
+ * the following procedure:
+ *
+ * * Find all compatible sub_tile types, based on which all potential sub_tile locations are found (this process is
+ *   made computationally cheap by legal_pos data member in AnalyticPlacer)
+ * * Within a radius (starting from 0) of the block's currently location, randomly choose a location as candidate.
+ * * If the block is a single block (not in macro), multiple candidates are potentially chosen, and the one that
+ *   results in the smallest input wirelength for the block is chosen.
```

Comment:
What is input wirelength?  wirelength to its input pins only?

Reply:
It's used in nextpnr, probably a heuristic. I did try adding up all wirelengths (input + output) and the result was slightly worse than input alone.

Reply:
Thanks.  Worth adding that to the comment briefly here (what is input wirelength), and adding why either here or in the strict legalizer's comments.